storage-proofs/src/piece_inclusion_proof.rs

```diff
@@ -1,255 +1,473 @@
+    fn subtree_capacity(pos: usize, total: usize) -> usize {
+        assert!(pos < total, "position must be less than tree capacity");
+
+        let mut capacity = 1;
+        // If tree is not 'full', then pos 0 will have subtree_capacity greater than size of tree.
+        let mut cursor = pos + next_pow2(total);
+
+        while cursor & 1 == 0 {
+            capacity *= 2;
+            cursor >>= 1;
         }
-        last_row = hash_pairs::<H>(&mut hasher, row.as_slice(), height)?;
+        capacity
     }
-    assert_eq!(last_row.len(), 1);
-    Ok(last_row[0])
 }
 
-/// For each successive pair of input hashes in row, construct a new hash according to the method used by `hasher.node`,
-/// returning a vector of the constructed hashes, in order.
-/// The result will be an error (resulting in a failed proof) if row does not contain an even number of hashes.
-fn hash_pairs<H: Hasher>(
-    hasher: &mut H::Function,
-    row: &[H::Domain],
-    height: usize,
-) -> Result<Vec<H::Domain>> {
-    let hashed: Result<Vec<_>> = row
-        .chunks(2)
-        .map(|pair| {
-            if pair.len() != 2 {
-                // If input is malformed, return an Err, which will fail the proof.
-                return Err(Error::MalformedInput);
-            }
-            hasher.reset();
-            Ok(hasher.node(pair[0], pair[1], height))
-        })
-        .collect();
+#[allow(dead_code)]
+fn create_piece_tree<H: Hasher>(
+    data: &[H::Domain],
+) -> merkle::MerkleTree<H::Domain, H::Function, VecStore<H::Domain>> {
+    let data_size = data.len();
+    // We need to compute comm_p as a merkle root over power-of-two-sized data.
+    let tree_size = next_pow2(data_size);
+
+    // If actual data is less than tree size, pad it with zeroes.
+    if data_size < tree_size {
+        let zero = H::Domain::try_from_bytes(&[0; 32]).unwrap();
+        let mut data_copy = vec![zero; tree_size];
+        data_copy[0..data_size].copy_from_slice(data);
+        merkle::MerkleTree::<H::Domain, H::Function, VecStore<H::Domain>>::new(data_copy)
+    } else {
+        merkle::MerkleTree::<H::Domain, H::Function, VecStore<H::Domain>>::new(data.to_vec())
+    }
+}
 
-    hashed
+#[allow(dead_code)]
+/// Compute comm_p from a slice of Domain elements
+fn compute_comm_p<H: Hasher>(data: &[H::Domain]) -> H::Domain {
+    create_piece_tree::<H>(data).root()
 }
 
-/// Return a vector of (hash, bool) pairs, where the bool indicates whether the paired hash
-/// will be the left child of the next node hashed. In order to accomplish this, we skip
-/// the first hash provided (in proof.lemma()). This is an implementation detail of how a
-/// merkle_light::proof::Proof is structured.
-fn proof_vec<T: Domain>(proof: &Proof<T>) -> impl Iterator<Item = (&T, &bool)> {
-    proof.lemma().iter().skip(1).zip(proof.path().iter())
+#[allow(dead_code)]
+/// Generate comm_p from bytes
+fn generate_comm_p<H: Hasher>(data: &[u8]) -> Result<H::Domain> {
```

Comment:
The naming seems odd, maybe `commitment_from_domain` and `commitment_from_bytes`, would make the intent of these methods clearer?